c/src/ml-api-service.c

```diff
@@ -0,0 +1,136 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file ml-api-service.c
+ * @date 07 Mar 2022
+ * @brief NNStreamer/Service C-API implementation
+ * @see	https://github.com/nnstreamer/nnstreamer
+ * @author MyungJoo Ham <myungjoo.ham@samsung.com>
+ * @author Sangjung Woo <sangjung.woo@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#include <nnstreamer.h>
+#include <leveldb/c.h>
+
+#include "ml-api-inference-internal.h"
+#include "ml-api-service.h"
+
+/* TODO: Need to set when configuration */
+#define SYS_DB_DIR            "./"
+#define ML_DATABASE_PATH      SYS_DB_DIR"/.ml-service-leveldb"
+
+static leveldb_t *ml_service_db;
+static leveldb_options_t *db_options;
+static leveldb_readoptions_t *db_roptions;
+static leveldb_writeoptions_t *db_woptions;
+
+static int service_init_db (void);
+
+int
+ml_service_pipeline_add (const char *name, const char *pipeline_desc)
+{
+  int ret = 0;
+  char *err = NULL;
+  gsize name_len;
+  gsize pipeline_len;
+
+  if (!name || !pipeline_desc) {
+    _ml_loge ("Error! name and pipeline_desc should not be NULL");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (!ml_service_db) {
+    ret = service_init_db ();
+    if (ret < 0)
+      return ret;
+  }
+
+  /* Put name and pipeline */
+  name_len = strlen (name);
+  pipeline_len = strlen (pipeline_desc);
+  leveldb_put (ml_service_db, db_woptions, name, name_len,
+      pipeline_desc, pipeline_len, &err);
+  if (err != NULL) {
+    _ml_loge ("Error! Failed to put the name: %s and value: %s", name,
+        pipeline_desc);
+    return -1;
+  }
+
+  return ML_ERROR_NONE;
+}
+
+int
+ml_service_pipeline_get (const char *name, char **pipeline_desc)
+{
+  int ret = 0;
+  gsize read_len;
+  gsize name_len;
+  char *err = NULL;
+  char *value = NULL;
+
+  if (!name || !*pipeline_desc) {
```

Comment:
Should be changed..?
```suggestion
  if (!name || !pipeline_desc) {
```

and if `*pipeline_desc` exists, need to free before allocate new memory.

Reply:
Right. I need to fix the code. 
By the way, do we need to free if the parameter pointer is allocated?

Reply:
If `ml_service_pipeline_get` is called twice, isn't it a memory leak?

Reply:
In that case, that is a memory leak. 
What about adding a @note section that allocated memory space should be freed by the caller?

Reply:
Yes, I think adding a note section is enough for this.